saltstack/base/salt/luks/bin/create-luks-volume.sh

```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
+
+set -ex -o pipefail
+
+: ${SECRET_ENCRYPTION_KEY_SOURCE:? required}
+
+export LUKS_BACKING_FILE_DEFAULT_SIZE="100MiB"
+
+if mountpoint /mnt/luks; then
+  echo "Volume already mounted at mount point... Exiting LUKS volume creation!"
+  exit
+fi
+
+if [[ -e /etc/luks/luks && $(stat -c "%a" /etc/luks/luks | tr -d '\n') == 600 ]]; then
+  # Set the backing file to the specified size if it exists
+  truncate -s "$LUKS_BACKING_FILE_DEFAULT_SIZE" /etc/luks/luks
+else
+  # Create the backing file if it doesn't exist
+  dd if=/dev/zero of=/etc/luks/luks bs="$LUKS_BACKING_FILE_DEFAULT_SIZE" count=1
+  chmod 600 /etc/luks/luks
+fi
+
+if ! mountpoint /mnt/luks_passphrase_tmpfs; then
+  # Create the tmpfs for the plaintext passphrase
+  mount -t tmpfs -o size=1k,mode=700 tmpfs /mnt/luks_passphrase_tmpfs
+fi
+
+if [[ -e /mnt/luks_passphrase_tmpfs/passphrase ]]; then
+  echo "Plaintext passphrase already exists... Exiting LUKS volume creation with failed exit code!"
+  exit 1
+else
+  aws kms generate-random \
+      --number-of-bytes 64 \
+      --output text \
+      --query Plaintext \
+    | base64 --decode \
+    1> /mnt/luks_passphrase_tmpfs/passphrase
+fi
+
+# Encrypt the plaintext passphrase and store the ciphertext in persistent storage
+if ! /usr/local/bin/aws-encryption-cli \
+       --encrypt \
+       --input /mnt/luks_passphrase_tmpfs/passphrase \
+       --output /etc/luks/passphrase_ciphertext \
+       --wrapping-keys provider=aws-kms key="$SECRET_ENCRYPTION_KEY_SOURCE" \
+       --metadata-output "/var/log/luks/passphrase_encryption_md-$(date +"%F-%T").json" \
```

Comment:
> Metadata Contents[](https://aws-encryption-sdk-cli.readthedocs.io/en/latest/#metadata-contents)
> The metadata JSON contains the following fields:
> - "mode" : "encrypt"/"decrypt"/"decrypt-unsigned"
> - "input" : Full path to input file (or "<stdin>" if stdin)
> - "output" : Full path to output file (or "<stdout>" if stdout)
> - "header" : JSON representation of [message header data](http://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/message-format.html#header-structure)
> - "header_auth" : JSON representation of [message header authentication data](http://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/message-format.html#header-authentication) (only on decrypt)

I don't think anything here is sensitive, but I'm not 100% sure.

Reply:
```json
{
  "header": {
    "algorithm": "AES_256_GCM_HKDF_SHA512_COMMIT_KEY_ECDSA_P384",
    "commitment_key": "VBM/Pw4HRf6jOI15qQEYUkCU0hRLCNheDh7m2OQ/KsU=",
    "content_type": 2,
    "encrypted_data_keys": [
      {
        "encrypted_data_key": "AQIBAHjNY1+vWYKur3ANO12lNTnqtta8fuEMpqkruayWZfi1BgFvdts9YF+1x0fm4ISB3uWRAAAAfjB8BgkqhkiG9w0BBwagbzBtAgEAMGgGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMUieGo91dbJ0h4M2bAgEQgDtdHaYG8Qbo8KgMe7xwQkgR9zkK4Vd9DCDtPSystBFZJlrMFsenRyPj2n8nPxcGbgvrTUiz66BM/38ncw==",
        "key_provider": {
          "key_info": "YXJuOmF3cy11cy1nb3Y6a21zOnVzLWdvdi13ZXN0LTE6MjAyNTUwMTY1MDcyOmtleS9mOWRlYTg4Yy1lMWM3LTQ3ZGUtOTZiMi03OWIxZTVlMjc1ZDU=",
          "provider_id": "YXdzLWttcw=="
        }
      }
    ],
    "encryption_context": {
      "INSTANCE_ID": "i-0f7902a80e0658a05",
      "aws-crypto-public-key": "Ah/B1SIX+0MrsarccEch7Zil8DJxQmFk953LYsCLgzijlUubQFYvxr90Zn3vhI/ULQ=="
    },
    "frame_length": 4096,
    "header_iv_length": null,
    "message_id": "RIKnSrpuUq74i+T72umNYIeyrk6H6R24FwXB3SXJ+HM=",
    "type": null,
    "version": "2.0"
  },
  "input": "/mnt/cdp-luks_passphrase_tmpfs/passphrase",
  "mode": "encrypt",
  "output": "/etc/cdp-luks/passphrase_ciphertext"
}
```
This is an example metadata JSON. I would need to check if any of this is sensitive, but on first look only the `encrypted_data_key` seems like something that could be sensitive.

Reply:
Also, seems like the `aws-encryption-sdk-cli` creates the file with `640` permissions by default. Is that restrictive enough, or would it need to be `600` for example?

Reply:
I would definitely go with `600` for these ones, for consistency. `encrypted_data_key` can be decrypted with the KMS key, and every cluster user has access to the latter (via the instance profile role Principal in the KMS key policy. Although that is only the data key itself; the passphrase plaintext or ciphertext are not accessible to regular users (only to `root`), so having the data key plaintext would not gain anything. :smile: 

That aside, `encrypted_data_key` is a ciphertext, so there is no harm revealing it in the metadata JSON (if restricted to `root` only), because the KMS key is kept outside the cluster / disk.

Reply:
`aws-crypto-public-key` seems to be only used for the AAD of the digital signature, so it is not sensitive either.